internal/autoevent/executor.go

```diff
@@ -47,11 +49,14 @@ func (e *executor) Run() {
 
 		if evt != nil {
 			if e.autoEvent.OnChange {
+				e.mutex.Lock()
 				if compareReadings(e, evt.Readings) {
 					common.LoggingClient.Debug(fmt.Sprintf("AutoEvent - readings are the same as previous one %v", e.lastReadings))
+					e.mutex.Unlock()
 					continue
 				}
 				cacheReadings(e, evt.Readings)
+				e.mutex.Unlock()
```

Comment:
@cloudxxx8 This whole thing is tricky because `executor.Run()` is called via a goroutine and you don't want the use of a Mutex to negate the concurrent usage of the executor. The way this is written, every read is included within the Mutex so only one goroutine can read at a time, effectively serializing access to the executor's `readings` map. If you're going to leave the code as is, I would suggest changing the type of the `e.mutex` variable to `sync.RWMutex`. See [here](https://golang.org/pkg/sync/#RWMutex). If you leave the type as a regular `Mutex`, you might want to do the following:

```
func (e *executor) Run() {
...
    if e.autoEvent.OnChange {
            if compareReadings(e, evt.Readings) {
                common.LoggingClient.Debug(fmt.Sprintf("AutoEvent - readings are the same as previous one %v", e.lastReadings))
                continue
        }
        // Notice the elimination of e.mutex from this function!
        cacheReadings(e, evt.Readings)
...
}

// Mutex moved here where the write happens, uses defer to ensure unlock
func cacheReadings(e *executor, readings []contract.Reading) {
	e.mutex.Lock()
	defer e.mutex.Unlock()
	for _, r := range readings {
		e.lastReadings[r.Name] = r.Value
	}
}
```
And actually, now that I've written this out, this method of code organization might help regardless of Mutex vs RWMutex, mostly because of the `defer` ensuring an unlock is called.

Reply:
@tsconn23 thanks for the instruction.
however, the code I wrote is not really working like you said.  because the mutex here is not the global variable across all the gorutine, but it is the local variable in each executor.  Thus, it won't lock others or block other gorutine.  I am thinking about how to optimize the code from your suggestion.

Reply:
please see the example I made:
https://play.golang.org/p/JmLvfMF8mFU

mutex block the gorutines each other when running this example.

By uncommeting line 18, the mutex becomes local variable in each gorutine, and they won't block each other.

Reply:
@cloudxxx8 OK, well you certainly know the SDK code better than I do so I'll defer to your judgment. Get it to where you think it needs to be and we'll merge it tomorrow so it's included in the release.

Reply:
thanks, your suggestion is better, so I am using RWMutex and move the lock to the function with defer unlock.  Please help review and confirm tomorrow morning.  I will keep awake until this issue resolved.